xivo/wsgi/__init__.py

```diff
@@ -0,0 +1,39 @@
+import logging
+import time
+
+from cheroot import wsgi
+
+logger = logging.getLogger(__name__)
+
+
+class HTTPServer_patch(wsgi.Server):
+    def serve(self) -> None:
+        """Serve requests, after invoking :func:`prepare()`."""
+        while self.ready and not self.interrupt:
+            try:
+                self._connections.run(self.expiration_interval)
+            except OSError:
+                self.error_log(
+                    'OS Error while serving an HTTP request',
+                    level=logging.ERROR,
+                    traceback=True,
+                )
+                exit(1)
+            except (KeyboardInterrupt, SystemExit):
+                raise
+            except Exception:
+                self.error_log(
+                    'Error in HTTPServer.serve',
+                    level=logging.ERROR,
+                    traceback=True,
+                )
+        # raise exceptions reported by any worker threads,
+        # such that the exception is raised from the serve() thread.
+        if self.interrupt:
+            while self._stopping_for_interrupt:
+                time.sleep(0.1)
+            if self.interrupt:
+                raise self.interrupt
+
+
+wsgi.WSGIServer.serve = HTTPServer_patch.serve
```

Comment:
Monkeypatching should be avoided, IMHO.

Since any project using cheroot would have to import this anyway, I don't think a monkey patch is required. 
The projects using cheroot should be refactored to use this patched class instead.

Reply:
For example, in the wazo-dird codebase, the `wsgi.WSGISServer` is used here:
https://github.com/wazo-platform/wazo-dird/blob/01d8d121396d82d1584496fa0072a9fa59043a6b/wazo_dird/http_server.py#L65-L69
The `from cheroot import wsgi` would be changed to `from xivo import wsgi`.

You can add an alias `WSGIServer = PatchedWSGIServer` such that only the imports need to change, and not the name of the class used.

Reply:
It's meant to be temporary, as I'm also planning on opening a bug ticket upstream. The current patch only need to be imported to work, then once the bug is fixed we can remove the import line

Reply:
Yes but we don't know how long upstream resolution will take.  
An additional import seems more error-prone, as removing the additional import(e.g. by mistake) will not produce any warning signs. Whereas if we substitute the existing import with the patched version, that cannot happen.

I see no advantage to the monkeypatch version here.